File: src/info/meodinger/LabelPlusFX/Util/CString.java
package info.meodinger.LabelPlusFX.Util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Meodinger
 * Date: 2021/5/29
 * Location: info.meodinger.LabelPlusFX.Util
 */
public class CString {

    public static char[] WHITE_SPACE_ARRAY = new char[] {
      ' ', '\0', '\b', '\f', '\n', '\r', '\t'
    };

    public static String repeat(String str, int n) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < n; i++) {
            builder.append(str);
        }
        return builder.toString();
    }

    public static String repeat(char c, int n) {
        return repeat(String.valueOf(c), n);
    }

    public static <T> int lengthOf(T t) {
        return String.valueOf(t).length();
    }

    public static boolean isWhiteSpace(char c) {
        for (char w : WHITE_SPACE_ARRAY) {
            if (c == w) return true;
        }
        return false;
    }

    public static boolean isBlank(String str) {
        if (str == null) return true;
        if (str.isEmpty()) return true;

        char[] chars = str.toCharArray();
        int whiteCount = 0;
        for (char c : chars) {
            if (isWhiteSpace(c)) whiteCount++;
        }
        return whiteCount == chars.length;
    }

    public static boolean isDigit(String str) {
        if (str == null) return false;

        char[] chars = str.toCharArray();
        for (char c : chars) {
            if (c < '0' || c > '9') return false;
        }
        return true;
    }

    public static List<String> trimSame(List<String> strings) {

        if (strings == null) return null;
        if (strings.isEmpty()) return Arrays.asList("", "");
        if (strings.size() == 1) return Arrays.asList("", "", strings.get(0));

        ArrayList<String> trimmed = new ArrayList<>();

        char[] example = strings.get(0).toCharArray();
        int head = example.length, tail = example.length;
        for (int i = 1; i < strings.size(); i++) {
            char[] chars = strings.get(i).toCharArray();
            int range = Math.min(example.length, chars.length);

            for (int j = 0; j < range; j++) {
                if (chars[j] != example[j]) {
                    head = Math.min(j, head);
                    break;
                }
            }
            for (int j = 0; j < range; j++) {
                if (chars[chars.length - j - 1] != example[example.length - j - 1]) {
                    tail = Math.min(j, tail);
                    break;
                }
            }
        }

        trimmed.add(strings.get(0).substring(0, head));
        trimmed.add(strings.get(0).substring(strings.get(0).length() - tail));
        for (String string : strings) trimmed.add(string.substring(head, string.length() - tail));

        return trimmed;
    }
}

File: src/main/kotlin/info/meodinger/lpfx/util/CString.kt
package info.meodinger.lpfx.util.string

import info.meodinger.lpfx.util.resource.I18N
import info.meodinger.lpfx.util.resource.get

import javafx.scene.text.Font
import javafx.scene.text.Text
import java.util.*


/**
 * Author: Meodinger
 * Date: 2021/7/29
 * Location: info.meodinger.lpfx.util
 */

/**
 * Delete tail string
 *
 * @param tail Tail to delete
 */
fun StringBuilder.deleteTail(tail: String) {
    if (tail.isEmpty()) return

    val tailLength = tail.length
    val builderLength = this.length

    this.deleteRange(builderLength - tailLength, builderLength)
}

/**
 * Is a string only contains 0-9
 */
fun String.isInt(): Boolean {
    val chars = this.toCharArray()
    for (c in chars) {
        if (c !in '0'..'9') return false
    }
    return true
}

/**
 * Trim a list of string
 *
 * @param strings Strings to trim
 * @return A list of String, first two is the same head & tail of input strings, others are strings that trimmed head & tail
 */
fun trimSame(strings: List<String>): List<String> {
    if (strings.isEmpty()) return listOf("", "")
    if (strings.size == 1) return listOf("", "", strings[0])

    val trimmed = ArrayList<String>()

    val example = strings[0].toCharArray()
    var head = example.size
    var tail = example.size
    for (i in 1 until strings.size) {
        val chars = strings[i].toCharArray()
        val range = example.size.coerceAtMost(chars.size)

        for (j in 0 until range) {
            if (chars[j] != example[j]) {
                head = head.coerceAtMost(j)
                break
            }
        }
        for (j in 0 until range) {
            if (chars[chars.size - j - 1] != example[example.size - j - 1]) {
                tail = tail.coerceAtMost(j)
                break
            }
        }
    }

    trimmed.add(strings[0].substring(0, head))
    trimmed.add(strings[0].substring(strings[0].length - tail))
    for (string in strings) trimmed.add(string.substring(head, string.length - tail))

    return trimmed
}

/**
 * Sort a list of string by their int value
 */
fun sortByDigit(strings: List<String>): List<String> {
    val trimmed = trimSame(strings)

    if (trimmed.size > 2) {
        var canCastToIntList = true
        for (i in 2 until trimmed.size) {
            if (!trimmed[i].isInt()) {
                canCastToIntList = false
                break
            }
        }
        if (canCastToIntList) {
            val sorted = trimmed.subList(2, trimmed.size).sortedBy { it.toInt() }

            return List(sorted.size) { trimmed[0] + sorted[it] + trimmed[1] }
        }
    }

    // default
    return strings.toMutableList().sorted()
}

/**
 * Omit high text
 */
fun omitHighText(longText: String, maxRowCount: Int = 10): String {
    val lines = longText.split("\n")

    if (lines.size <= maxRowCount) return longText

    val builder = StringBuilder()
    for (i in 0 until maxRowCount - 1) builder.appendLine(lines[i])
    builder.append(String.format(I18N["util.long_text.format.i"], lines.size - maxRowCount - 1))

    return builder.toString()
}

/**
 * Omit wide text
 */
fun omitWideText(longText: String, maxWidth: Double, font: Font? = null): String {
    val lines = longText.split("\n")
    val builder = StringBuilder()

    val t = Text().also { it.font = font }
    val b = StringBuilder()
    var p: Int
    for (line in lines) {
        t.text = line
        b.clear()
        p = t.text.length - 1
        if (t.boundsInLocal.width > maxWidth) {
            b.append(line).append("...")
            while (t.boundsInLocal.width > maxWidth) {
                b.deleteAt(p--)
                t.text = b.toString()
            }
            builder.appendLine(b.toString())
        } else {
            builder.appendLine(line)
        }
    }
    if (builder.isNotEmpty()) builder.deleteTail("\n")

    return builder.toString()
}